src/backend/executor/nodeHash.c

```diff
@@ -3769,7 +3816,12 @@ ExecHashTableDetachBatch(HashJoinTable hashtable)
 		sts_end_parallel_scan(hashtable->batches[curbatch].outer_tuples);
 
 		/* Detach from the batch we were last working on. */
-		if (BarrierArriveAndDetach(&batch->batch_barrier))
+		/*
+		 * CBDB_PARALLEL: Parallel Hash Left Anti Semi (Not-In) Join(parallel-aware)
+		 * If phs_lasj_has_null is true, that means we have found null when building hash table,
+		 * there were no batches to detach.
+		 */
+		if (!hashtable->parallel_state->phs_lasj_has_null && BarrierArriveAndDetach(&batch->batch_barrier))
```

Comment:
No need to Detach in some case ? Consider `Detach` always ?

Reply:
> No need to Detach in some case ?

Yeah, as comments: when  phs_lasj_has_null is true

> Consider Detach always ?

Can't do that, detach without attach will hang forever.

Reply:
since phs_lasj_has_null is not atomic , `BarrierArriveAndDetach ` maybe called when another process set phs_lasj_has_null to true, but `phs_lasj_has_null` in this process is false.

Reply:
> BarrierArriveAndDetach  maybe called when another process set phs_lasj_has_null to true

Impossible, if someone is setting phs_lasj_has_null to true, all processes are in building hash inner phase, can't arrive here.

Reply:
> since phs_lasj_has_null is not atomic , BarrierArriveAndDetach  maybe called when another process set phs_lasj_has_null to true, but phs_lasj_has_null in this process is false.

After a think: phs_lasj_has_null is atomic, but it's possible for such a extreme case:  a process doesn't participate in building hash inner and doesn't get the real-time phs_lasj_has_null though others have set it to true in shared memory( CPU L1 cache coherency).

A `pg_memory_barrier` in  MultiExecParallelHash will prevent that, I will add it, thanks.

Reply:
> A pg_memory_barrier in MultiExecParallelHash will prevent that, I will add it, thanks.

Done.